File: python/xenarch/webhooks.py
"""Webhook signature verification for merchant + agent-operator backends.

Every pay-link and every agent can POST events (payment.confirmed,
subscription.renewed, cap.exceeded, scope.denied, ...) to your server. Each
delivery carries a Stripe-style timestamped signature::

    X-Xenarch-Signature: t=<unix_seconds>,v1=<hex>
        where v1 = HMAC_SHA256(secret, "<unix_seconds>.<raw_body>")
    X-Xenarch-Event:     <event_type>     e.g. payment.confirmed
    X-Xenarch-Delivery:  <uuid>           idempotency key for retries

Signing the timestamp alongside the body makes a captured delivery
un-replayable: ``verify`` rejects any request whose ``t`` is more than
``tolerance_seconds`` (default 300) from now. The same scheme covers both
pay-link and agent webhooks — one ``verify`` call works for either.

    from xenarch import webhooks

    @app.post("/xenarch-webhook")
    async def hook(request):
        body = await request.body()
        sig = request.headers["X-Xenarch-Signature"]
        if not webhooks.verify(body, sig, secret):
            return Response(status_code=401)
        event = json.loads(body)
        ...
"""
from __future__ import annotations

import hashlib
import hmac
import time

__all__ = [
    "verify",
    "compute_signature",
    "WebhookVerificationError",
    "DEFAULT_TOLERANCE_SECONDS",
]

# Replay window: reject deliveries whose signed ``t`` is older/newer than this.
DEFAULT_TOLERANCE_SECONDS = 300


class WebhookVerificationError(Exception):
    """Raised by ``verify(..., raise_on_failure=True)`` on a bad signature."""


def _signed_bytes(timestamp: int, body: bytes) -> bytes:
    return f"{timestamp}.".encode("utf-8") + body


def _hmac_hex(message: bytes, secret: str) -> str:
    return hmac.new(
        secret.encode("utf-8"), msg=message, digestmod=hashlib.sha256
    ).hexdigest()


def compute_signature(payload: bytes | str, secret: str, timestamp: int) -> str:
    """Full ``X-Xenarch-Signature`` value for ``payload`` at ``timestamp``.

    Returns ``t=<ts>,v1=<hex>`` — byte-for-byte what the platform sends.
    """
    body = payload.encode("utf-8") if isinstance(payload, str) else payload
    hex_sig = _hmac_hex(_signed_bytes(timestamp, body), secret)
    return f"t={timestamp},v1={hex_sig}"


def _parse_header(header: str) -> tuple[int, str] | None:
    """Parse ``t=<int>,v1=<hex>`` (order-independent). None if malformed."""
    t: int | None = None
    v1: str | None = None
    for part in header.strip().split(","):
        key, sep, val = part.partition("=")
        if not sep:
            continue
        key = key.strip()
        val = val.strip()
        if key == "t":
            try:
                t = int(val)
            except ValueError:
                return None
        elif key == "v1":
            v1 = val
    if t is None or v1 is None:
        return None
    return t, v1


def verify(
    payload: bytes | str,
    signature_header: str,
    secret: str,
    *,
    tolerance_seconds: int = DEFAULT_TOLERANCE_SECONDS,
    now: int | None = None,
    raise_on_failure: bool = False,
) -> bool:
    """Verify a webhook signature in constant time, with replay protection.

    Args:
        payload: the exact raw request body bytes (do not re-serialize a parsed
            dict — re-serialization can change bytes and break the signature).
            A ``str`` is UTF-8 encoded.
        signature_header: the ``X-Xenarch-Signature`` value, e.g. ``t=...,v1=...``.
        secret: the link's or agent's ``whsec_...`` webhook secret.
        tolerance_seconds: replay window (default 300).
        now: override "now" (unix seconds) — for tests.
        raise_on_failure: raise ``WebhookVerificationError`` instead of
            returning ``False``.

    Returns True when the signature matches and the timestamp is fresh.
    """

    def _fail(msg: str) -> bool:
        if raise_on_failure:
            raise WebhookVerificationError(msg)
        return False

    # Fail closed on a missing/empty header — the common misuse.
    if not signature_header:
        return _fail("missing webhook signature header")

    parsed = _parse_header(signature_header)
    if parsed is None:
        return _fail("malformed webhook signature header")
    timestamp, v1 = parsed

    current = int(time.time()) if now is None else now
    if abs(current - timestamp) > tolerance_seconds:
        return _fail("webhook timestamp outside tolerance (replay?)")

    body = payload.encode("utf-8") if isinstance(payload, str) else payload
    expected = _hmac_hex(_signed_bytes(timestamp, body), secret)
    if not hmac.compare_digest(expected, v1):
        return _fail("webhook signature does not match")
    return True

File: js/sdk/src/webhooks.ts
/**
 * Webhook signature verification for merchant + agent-operator backends.
 *
 * Every pay-link and every agent can POST events (payment.confirmed,
 * subscription.renewed, cap.exceeded, scope.denied, ...) to your server. Each
 * delivery carries a Stripe-style timestamped signature:
 *
 *     X-Xenarch-Signature: t=<unix_seconds>,v1=<hex>
 *         where v1 = HMAC_SHA256(secret, "<unix_seconds>.<raw_body>")
 *     X-Xenarch-Event:     <event_type>     e.g. payment.confirmed
 *     X-Xenarch-Delivery:  <uuid>           idempotency key for retries
 *
 * Signing the timestamp alongside the body makes a captured delivery
 * un-replayable: `verify` rejects any request whose `t` is more than
 * `toleranceSeconds` (default 300) from now. Verify before trusting the
 * payload, the same way you'd verify a Stripe webhook. The same scheme covers
 * both pay-link and agent webhooks — one `verify` call works for either.
 *
 * Isomorphic: uses Web Crypto (`crypto.subtle`), so it runs on Node, Bun,
 * Deno, and edge runtimes (Cloudflare Workers, Vercel Edge).
 *
 *     import { webhooks } from "@xenarch/sdk";
 *
 *     const raw = await request.text();           // raw body, not a parsed object
 *     const sig = request.headers.get("X-Xenarch-Signature");
 *     if (!(await webhooks.verify(raw, sig, secret))) {
 *       return new Response("bad signature", { status: 401 });
 *     }
 *     const event = JSON.parse(raw);
 */
import { WebhookVerificationError } from "./errors.js";

const encoder = new TextEncoder();

/** Replay window: reject deliveries whose signed `t` is older/newer than this. */
export const DEFAULT_TOLERANCE_SECONDS = 300;

function toBytes(payload: string | Uint8Array): Uint8Array {
  return typeof payload === "string" ? encoder.encode(payload) : payload;
}

/** Concatenate `"<timestamp>."` + body into the exact bytes that get signed. */
function signedBytes(timestamp: number, payload: string | Uint8Array): Uint8Array {
  const prefix = encoder.encode(`${timestamp}.`);
  const body = toBytes(payload);
  const out = new Uint8Array(prefix.length + body.length);
  out.set(prefix, 0);
  out.set(body, prefix.length);
  return out;
}

async function hmacHex(message: Uint8Array, secret: string): Promise<string> {
  const key = await crypto.subtle.importKey(
    "raw",
    encoder.encode(secret) as unknown as BufferSource,
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  // `as unknown as BufferSource`: TS types Uint8Array as generic over its
  // backing buffer, which the DOM `BufferSource` doesn't accept directly.
  const sig = await crypto.subtle.sign(
    "HMAC",
    key,
    message as unknown as BufferSource,
  );
  return [...new Uint8Array(sig)]
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

/**
 * Compute the full `X-Xenarch-Signature` header value for `payload` at
 * `timestamp` (unix seconds), keyed by `secret`. Returns `t=<ts>,v1=<hex>` —
 * byte-for-byte what the platform sends.
 */
export async function computeSignature(
  payload: string | Uint8Array,
  secret: string,
  timestamp: number,
): Promise<string> {
  const hex = await hmacHex(signedBytes(timestamp, payload), secret);
  return `t=${timestamp},v1=${hex}`;
}

/** Parse `t=<int>,v1=<hex>` (order-independent). Returns null if malformed. */
function parseSignatureHeader(
  header: string,
): { t: number; v1: string } | null {
  let t: number | null = null;
  let v1: string | null = null;
  for (const part of header.trim().split(",")) {
    const eq = part.indexOf("=");
    if (eq === -1) continue;
    const k = part.slice(0, eq).trim();
    const val = part.slice(eq + 1).trim();
    if (k === "t") {
      const parsed = Number.parseInt(val, 10);
      if (Number.isFinite(parsed)) t = parsed;
    } else if (k === "v1") {
      v1 = val;
    }
  }
  if (t === null || v1 === null) return null;
  return { t, v1 };
}

export interface VerifyOptions {
  /** Throw `WebhookVerificationError` instead of returning `false`. */
  throwOnFailure?: boolean;
  /** Replay window in seconds (default 300). */
  toleranceSeconds?: number;
  /** Override "now" (unix seconds) — for tests. */
  now?: number;
}

/**
 * Verify a webhook signature in constant time, with replay protection.
 *
 * @param payload the exact raw request body (do not re-serialize a parsed
 *   object — re-serialization can change bytes and break the signature).
 * @param signatureHeader the `X-Xenarch-Signature` value, e.g. `t=...,v1=...`.
 * @param secret the link's or agent's `whsec_...` webhook secret.
 */
export async function verify(
  payload: string | Uint8Array,
  signatureHeader: string | null | undefined,
  secret: string,
  opts: VerifyOptions = {},
): Promise<boolean> {
  const fail = (msg: string): boolean => {
    if (opts.throwOnFailure) throw new WebhookVerificationError(msg);
    return false;
  };

  // Fail closed on a missing header — the common misuse and an attacker's
  // first move.
  if (!signatureHeader) return fail("missing webhook signature header");

  const parsed = parseSignatureHeader(signatureHeader);
  if (!parsed) return fail("malformed webhook signature header");

  const tolerance = opts.toleranceSeconds ?? DEFAULT_TOLERANCE_SECONDS;
  const now = opts.now ?? Math.floor(Date.now() / 1000);
  if (Math.abs(now - parsed.t) > tolerance) {
    return fail("webhook timestamp outside tolerance (replay?)");
  }

  const expected = await hmacHex(signedBytes(parsed.t, payload), secret);
  if (!timingSafeEqual(expected, parsed.v1)) {
    return fail("webhook signature does not match");
  }
  return true;
}

/** Length-checked constant-time string compare (no early-exit on mismatch). */
function timingSafeEqual(a: string, b: string): boolean {
  if (a.length !== b.length) return false;
  let result = 0;
  for (let i = 0; i < a.length; i++) {
    result |= a.charCodeAt(i) ^ b.charCodeAt(i);
  }
  return result === 0;
}

export const webhooks = { verify, computeSignature, DEFAULT_TOLERANCE_SECONDS };
